mingw doesn't have gmtime_s()

### Description
- Type: Bug
- Priority: Blocker

---------------------------------------------------------------
## Bug

**OS**  
linux|windows

**mbed TLS build:**  
Version: 0592ea772aee48ca1e6d9eb84eca8e143033d973

Building with mingw32, with the default configuration fails because mingw doesn't have `gmtime_s()`.

Extract from the build:
```
"  CC    platform.c"
"  CC    platform_util.c"
platform_util.c: In function 'mbedtls_platform_gmtime_r':
platform_util.c:109:15: warning: implicit declaration of function 'gmtime_s' [-W
implicit-function-declaration]
     return( ( gmtime_s( tm_buf, tt ) == 0 ) ? tm_buf : NULL );
               ^~~~~~~~
"  CC    poly1305.c"
"  CC    ripemd160.c"
"  CC    rsa_internal.c"
"  CC    rsa.c"
"  CC    sha1.c"
"  CC    sha256.c"
"  CC    sha512.c"
"  CC    threading.c"
"  CC    timing.c"
"  CC    version.c"
"  CC    version_features.c"
"  CC    xtea.c"
"  AR    libmbedcrypto.a"
"  CC    certs.c"
"  CC    pkcs11.c"
"  CC    x509.c"
"  CC    x509_create.c"
"  CC    x509_crl.c"
"  CC    x509_crt.c"
"  CC    x509_csr.c"
"  CC    x509write_crt.c"
"  CC    x509write_csr.c"
"  AR    libmbedx509.a"
"  CC    debug.c"
"  CC    net_sockets.c"
"  CC    ssl_cache.c"
"  CC    ssl_ciphersuites.c"
"  CC    ssl_cli.c"
"  CC    ssl_cookie.c"
"  CC    ssl_srv.c"
"  CC    ssl_ticket.c"
"  CC    ssl_tls.c"
"  AR    libmbedtls.a"
"  Gen   test_suite_x509parse.c"
"  CC    test_suite_x509parse.c"
../library/libmbedcrypto.a(platform_util.o):platform_util.c:(.text+0x44): undefi
ned reference to `gmtime_s'
collect2.exe: error: ld returned 1 exit status
Makefile:102: recipe for target 'test_suite_x509parse.exe' failed
mingw32-make[1]: *** [test_suite_x509parse.exe] Error 1
Makefile:21: recipe for target 'tests' failed
mingw32-make: *** [tests] Error 2
```

Mingw version:
```
C:\mbedtls\>gcc --version
gcc (MinGW.org GCC-6.3.0-1) 6.3.0
Copyright (C) 2016 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
```

The build should work out of the box (or tarball).

I have downloaded mingw64 version 8.1.0 for architecture x86_64, and with the `make CC=x86_64-w64-mingw32-gcc` and it seems that the `gmtime_s` issue is resolved, but got the following link errors:
~~~
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x7a): undefined reference to `__imp_getaddrinfo'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x98): undefined reference to `__imp_socket'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0xa5): undefined reference to `__imp_connect'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0xac): undefined reference to `__imp_closesocket'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x103): undefined reference to `__imp_freeaddrinfo'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x137): undefined reference to `__imp_WSAStartup'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x1e6): undefined reference to `__imp_getaddrinfo'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x203): undefined reference to `__imp_socket'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x20a): undefined reference to `__imp_setsockopt'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x211): undefined reference to `__imp_bind'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x218): undefined reference to `__imp_listen'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x297): undefined reference to `__imp_freeaddrinfo'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x2ca): undefined reference to `__imp_closesocket'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x2e4): undefined reference to `__imp_WSAStartup'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x315): undefined reference to `__imp_closesocket'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x325): undefined reference to `__imp_closesocket'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x39e): undefined reference to `__imp_getsockopt'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x3fa): undefined reference to `__imp_recvfrom'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x465): undefined reference to `__imp_connect'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x4a2): undefined reference to `__imp_accept'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x4b4): undefined reference to `__imp_WSAGetLastError'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x4f7): undefined reference to `__imp_getsockname'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x518): undefined reference to `__imp_socket'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x53f): undefined reference to `__imp_setsockopt'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x554): undefined reference to `__imp_bind'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x5a3): undefined reference to `__imp_WSAGetLastError'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x5eb): undefined reference to `__imp_ioctlsocket'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x61b): undefined reference to `__imp_ioctlsocket'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x692): undefined reference to `__imp_select'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x707): undefined reference to `__WSAFDIsSet'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x717): undefined reference to `__WSAFDIsSet'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x7b1): undefined reference to `__imp_recv'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x7c3): undefined reference to `__imp_WSAGetLastError'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x879): undefined reference to `__imp_select'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x8a2): undefined reference to `__imp_WSAGetLastError'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x8f1): undefined reference to `__imp_send'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x903): undefined reference to `__imp_WSAGetLastError'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x957): undefined reference to `__imp_shutdown'
../library/libmbedtls.a(net_sockets.o):net_sockets.c:(.text+0x960): undefined reference to `__imp_closesocket'
~~~
Need further investigation, might be missing winsock library

--- a/ChangeLog.d/mingw.txt
+++ b/ChangeLog.d/mingw.txt
@@ -1,0 +1,4 @@
+Bugfix
+   * Fix compilation error with mingw32. Fixed by Cameron Cawley in #4211.
+   * Fix compilation error when using C++ Builder on Windows. Reported by
+     Miroslav Mastny in #4015.

--- a/library/platform_util.c
+++ b/library/platform_util.c
@@ -92,9 +92,11 @@ void mbedtls_platform_zeroize( void *buf, size_t len )
  * threading.h. However, this macro is not part of the Mbed TLS public API, so
  * we keep it private by only defining it in this file
  */
-#if ! ( defined(_WIN32) && !defined(EFIX64) && !defined(EFI32) )
+#if ! ( defined(_WIN32) && !defined(EFIX64) && !defined(EFI32) ) || \
+      ( defined(__MINGW32__) && !defined(__MINGW64_VERSION_MAJOR) )
 #define PLATFORM_UTIL_USE_GMTIME
-#endif /* ! ( defined(_WIN32) && !defined(EFIX64) && !defined(EFI32) ) */
+#endif /* ! ( defined(_WIN32) && !defined(EFIX64) && !defined(EFI32) ) || \
+             (defined(__MINGW32__) && !defined(__MINGW64_VERSION_MAJOR) ) */
 
 #endif /* !( ( defined(_POSIX_VERSION) && _POSIX_VERSION >= 200809L ) ||     \
              ( defined(_POSIX_THREAD_SAFE_FUNCTIONS ) &&                     \
@@ -103,7 +105,7 @@ void mbedtls_platform_zeroize( void *buf, size_t len )
 struct tm *mbedtls_platform_gmtime_r( const mbedtls_time_t *tt,
                                       struct tm *tm_buf )
 {
-#if defined(_WIN32) && !defined(EFIX64) && !defined(EFI32)
+#if defined(_WIN32) && !defined(PLATFORM_UTIL_USE_GMTIME)
     return( ( gmtime_s( tm_buf, tt ) == 0 ) ? tm_buf : NULL );
 #elif !defined(PLATFORM_UTIL_USE_GMTIME)
     return( gmtime_r( tt, tm_buf ) );

--- a/library/platform_util.c
+++ b/library/platform_util.c
@@ -74,6 +74,7 @@ void mbedtls_platform_zeroize( void *buf, size_t len )
 #endif /* MBEDTLS_PLATFORM_ZEROIZE_ALT */
 
 #if defined(MBEDTLS_HAVE_TIME_DATE) && !defined(MBEDTLS_PLATFORM_GMTIME_R_ALT)
+#define __STDC_WANT_LIB_EXT1__ 1
 #include <time.h>
 #if !defined(_WIN32) && (defined(unix) || \
     defined(__unix) || defined(__unix__) || (defined(__APPLE__) && \
@@ -105,7 +106,12 @@ struct tm *mbedtls_platform_gmtime_r( const mbedtls_time_t *tt,
                                       struct tm *tm_buf )
 {
 #if defined(_WIN32) && !defined(PLATFORM_UTIL_USE_GMTIME)
+#if defined(__STDC_LIB_EXT1__)
+    return( ( gmtime_s( tt, tm_buf ) == 0 ) ? NULL : tm_buf );
+#else
+    /* MSVC and mingw64 argument order and return value are inconsistent with the C11 standard */
     return( ( gmtime_s( tm_buf, tt ) == 0 ) ? tm_buf : NULL );
+#endif
 #elif !defined(PLATFORM_UTIL_USE_GMTIME)
     return( gmtime_r( tt, tm_buf ) );
 #else

--- a/library/platform_util.c
+++ b/library/platform_util.c
@@ -95,8 +95,7 @@ void mbedtls_platform_zeroize( void *buf, size_t len )
 #if ! ( defined(_WIN32) && !defined(EFIX64) && !defined(EFI32) ) || \
       ( defined(__MINGW32__) && !defined(__MINGW64_VERSION_MAJOR) )
 #define PLATFORM_UTIL_USE_GMTIME
-#endif /* ! ( defined(_WIN32) && !defined(EFIX64) && !defined(EFI32) ) || \
-             (defined(__MINGW32__) && !defined(__MINGW64_VERSION_MAJOR) ) */
+#endif
 
 #endif /* !( ( defined(_POSIX_VERSION) && _POSIX_VERSION >= 200809L ) ||     \
              ( defined(_POSIX_THREAD_SAFE_FUNCTIONS ) &&                     \

--- a/library/platform_util.c
+++ b/library/platform_util.c
@@ -74,7 +74,7 @@ void mbedtls_platform_zeroize( void *buf, size_t len )
 #endif /* MBEDTLS_PLATFORM_ZEROIZE_ALT */
 
 #if defined(MBEDTLS_HAVE_TIME_DATE) && !defined(MBEDTLS_PLATFORM_GMTIME_R_ALT)
-#define __STDC_WANT_LIB_EXT1__ 1
+#define __STDC_WANT_LIB_EXT1__ 1  /* Ask for the C11 gmtime_s() if it's available */
 #include <time.h>
 #if !defined(_WIN32) && (defined(unix) || \
     defined(__unix) || defined(__unix__) || (defined(__APPLE__) && \

--- a/tests/suites/helpers.function
+++ b/tests/suites/helpers.function
@@ -44,9 +44,12 @@ typedef UINT32 uint32_t;
 
 #include <string.h>
 
+#if defined(__unix__) || (defined(__APPLE__) && defined(__MACH__)) || defined(__MINGW32__)
+#include <strings.h>
+#endif
+
 #if defined(__unix__) || (defined(__APPLE__) && defined(__MACH__))
 #include <unistd.h>
-#include <strings.h>
 #endif
 
 /* Type for Hex parameters */

